Fixes OWASP/mastg#2589 [MASWE-0025] Weak Signature (#4)
OWASP/maswe: weaknesses/MASVS-CRYPTO/MASWE-0025.md

Author: Sorry, could you clarify.

Reviewer: I think "Improper Generation of Digital Signatures" is the most accurate one. 

One way this can happen (be introduced; see Modes of Introduction) is due to "Using a deprecated, risky, or broken hashing algorithm"

---
title: Improper Generation of Digital Signatures
id: MASWE-0025
alias: improper-signature-generation

> platform: [android, ios]

Author: Refering to it can’t hurt.

---

> title: Signatures Created by Deprecated, Risky or Broken Algorithms

Reviewer: ```suggestion
title: Improper Generation of Cryptographic Signatures
```

id: MASWE-0025

> alias: weak-signatures

Reviewer: ```suggestion
alias: improper-signature-generation
```

- https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf
- https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf
- https://csrc.nist.gov/pubs/ir/8547/ipd
status: new

---

## Overview

Using deprecated, risky, or broken algorithms in signatures (such as MD5 or SHA-1) could enable an attacker to commit digital signature forgery, thereby undermining integrity, authenticity, and non-repudiation.

## Impact

> - **Loss of authenticity**: Signature forgery may allow the attacker to sign the data on behalf of another entity, thereby casting doubt on where the data came from.

Reviewer: ```suggestion
- **Loss of authenticity**: Signature forgery may allow an attacker to sign data on behalf of another entity, so the app can no longer trust where the data originated.
```

> - **Loss of accountability**: Signature forgery allows for plausible deniability and diminishes accountability.

Reviewer 2: To have accountability, we usually need identification, plus logging of actions. Non-repudiation seems more to the point here.

```suggestion
- **Loss of non-repudiation**: Signature forgery allows for plausible deniability and diminishes accountability.
```

> - **Loss of accountability**: Signature forgery allows for plausible deniability and diminishes accountability.

Reviewer 2: 1. I would add confidentiality: asymmetric algos also ensure it
2. I would structure the existing into: integrity (that data being accessed is intact and as expected) and non-repudiation (authenticity - that an action came from a user). Accountability is closer to logging users' actions, indeed, to place.

Author: To me, the Parkerian Hexad is more relevant in a larger system. For mobile, keep it simple and practical. We should focus on what is relevant within the context of mobile development and the use of signatures within that context. This way, it's easier for users of the MASWE, developers, to understand what is meant.

Author: @Diolor, please tell us if you disagree; otherwise, we may close the discussion.

- **Loss of accountability**: Signature forgery allows for plausible deniability and diminishes accountability.

## Modes of Introduction

- **Using a deprecated, risky, or broken hashing algorithm**: e.g., MD5 and SHA-1 have been identified to be vulnerable to collision attacks that are faster than a birthday attack. Because of this, they are denounced as "broken".

> - **Using an insufficiently collision-resistant hash**: Choosing a hashing algorithm of insufficient length may result in loss of integrity or confidentiality.

Reviewer: ```suggestion
- **Using deprecated, risky, or broken algorithms**: For example, choosing RSA SHA1 or RSA MD5 algorithms, even when documentation marks them as deprecated.
- **Using insufficient key sizes**: For example, generating RSA keys of 1024 bits or using non standard elliptic curves below current security levels.
- **Using insufficiently collision-resistant hash functions**: Selecting MD5 or SHA1, which have known collision attacks, can allow attackers to craft different messages with the same digest.
```

## Mitigations

- **Choose a collision-resistant algorithm**: Choose a signature algorithm that is sufficiently collision-resistant, like RSA (3072 bits and higher), ECDSA with NIST P-384, or EdDSA with Edwards448.

> - **Choose a signing scheme that makes use of algorithms with sufficient bit-lengths**: As our computers get stronger, previously generated hashes get weaker. Therefore, make sure you can adjust the bit length (strength) of the algorithm you choose. When signatures are stored at rest, make sure to follow the software industry's long-term recommendations (e.g., ["NIST: Transition to Post-Quantum Cryptography Standards"](https://csrc.nist.gov/pubs/ir/8547/ipd)).

Reviewer: ```suggestion
- **Choose a modern, collision-resistant algorithm with sufficient bit-lengths**: Choose a signature algorithm with at least 128 bits of security. Suitable choices include RSA with 3072 bit keys, ECDSA with NIST P256 or P384, and EdDSA with Ed25519 or Edwards448.
- Design signatures with algorithm agility. Include metadata or version fields that allow migration to stronger algorithms as recommended in documents like NIST SP 800-131A and post quantum transition guidance (["NIST: Transition to Post-Quantum Cryptography Standards"](https://csrc.nist.gov/pubs/ir/8547/ipd)).
- **Do not rely on provider defaults**: Enforce minimum algorithm and key requirements **explicitly** in the code.
```

> draft:
>   description: The use of algorithms with insufficient strength for signatures such as SHA1withRSA, etc. in a security-sensitive context should be avoided to ensure the integrity and authenticity of the data.
>   topics: null

Reviewer: ```suggestion
```

> status: draft
> 
> ---

Reviewer: ```suggestion
status: new

---

## Overview

```
